Create JSON Schema for YAML File

mkdjr opened the pull request:
Co-author: @jassielof

I took a pass at creating a JSON schema based on the [Hayagriva specification](https://github.com/typst/hayagriva/blob/main/docs/file-format.md). This will likely close https://github.com/typst/hayagriva/issues/33!

Please provide any and all feedback! This is my first time creating a JSON schema.

This PR might be related to issue https://github.com/typst/hayagriva/issues/134.



## Outstanding Items (copied from below for convenience)

- [ ] Are the tests in `tests/schema.rs` compliant with the package's architecture?
- [ ] Does the `title` field still accept a `translator` sub-field (see: https://github.com/typst/hayagriva/pull/192#discussion_r1966948068)?
- [ ] Upload to the [JSON Schema Store](https://www.schemastore.org/json/), as suggested by @Freed-Wu. See https://github.com/SchemaStore/schemastore/blob/master/CONTRIBUTING.md for more information.

mkdjr committed:
create json schema

mkdjr committed:
add examples from specification using the `examples` keyword

mkdjr committed:
use `$dynamicAnchor` and `$dynamicRef` according to the JSON 2020-12 schema

mkdjr committed:
update `parent` schema to correctly reference `#entry` if in an array, move description to root of `parent` schema

EpicEricEE commented:
The specification is not fully up to date as the fields added in #102 (`abstract`, `annote` and `genre`) aren't included and are thus also missing here.

mkdjr committed:
add `abstract`, `annote`, and `genre` to schema

MKDJr commented:
> The specification is not fully up to date as the fields added in #102 (`abstract`, `annote` and `genre`) aren't included and are thus also missing here.

Thanks for pointing this out! I added them to the schema and started a PR to add them to the documentation (https://github.com/typst/hayagriva/pull/194).

mkdjr committed:
add `annote` example to JSON schema

reknih commented on hayagriva.schema.json:
```diff
@@ -0,0 +1,436 @@
+    "title": "Hayagriva Bibliography Format",
+    "description": "A bibliography management format for the modern age.",
+    "type": "object",
+    "additionalProperties": false,
```
This blocks `patternProperties` from being used.

reknih commented on hayagriva.schema.json:
```diff
@@ -0,0 +1,436 @@
+    "title": "Hayagriva Bibliography Format",
+    "description": "A bibliography management format for the modern age.",
+    "type": "object",
+    "additionalProperties": false,
```
This blocks `patternProperties` from being used.

mkdjr replied:
`patternProperties` actually still works even when additional properties aren't allowed! ([source](https://json-schema.org/understanding-json-schema/reference/object#additionalproperties))

It should just prevent any entries that don't match the `patternProperties` regex.

reknih commented on hayagriva.schema.json:
```diff
@@ -0,0 +1,436 @@
+                        "video"
+                    ],
+                    "enum": [
+                        "article",
+                        "chapter",
+                        "entry",
+                        "anthos",
+                        "report",
+                        "thesis",
+                        "web",
+                        "scene",
+                        "artwork",
+                        "patent",
+                        "case",
+                        "newspaper",
+                        "legislation",
+                        "manuscript",
+                        "original",
+                        "post",
+                        "misc",
+                        "performance",
+                        "periodical",
+                        "proceedings",
+                        "book",
+                        "blog",
+                        "reference",
+                        "conference",
+                        "anthology",
+                        "repository",
+                        "thread",
+                        "video",
+                        "audio",
+                        "exhibition"
```
Types are case-insensitive. In particular, they can start with an uppercase letter. I don't think that there is a particularly clean solution to encoding this right now though: https://stackoverflow.com/questions/26604730/can-json-schema-enums-be-case-insensitive

mkdjr replied:
I took the brute force approach and converted each possible type into a RegEx pattern.
```
...
"oneOf": [
    {
        "pattern": "[Aa][Rr][Tt][Ii][Cc][Ll][Ee]"
    },
    ...
],
...
```

reknih commented on hayagriva.schema.json:
```diff
@@ -0,0 +1,436 @@
+                    ],
+                    "description": "The media type of the item. Often determines the structure of references."
+                },
+                "title": {
+                    "type": "string",
+                    "description": "The title of the item.",
+                    "examples": [
+                        "Rick Astley: How An Internet Joke Revived My Career"
+                    ]
+                },
```
Does not support the object mode:

```yaml
    title:
        value: "ZygOS: Achieving Low Tail Latency for Microsecond-Scale Networked Tasks"
        verbatim: true
        # also possible: sentence-case, title-case, short
```

mkdjr replied:
Do `sentence-case` and `title-case` still need to be handled? I just saw that they were [removed for any formattable string in `0.4.0`](https://github.com/typst/hayagriva/releases/tag/v0.4.0), is that still the case?

If so, `basic.yml` might need updating, as it lists certain `title` entries with those fields (https://github.com/typst/hayagriva/blob/edec836b302687316f9e5461e2bb133a326a6365/tests/data/basic.yml#L42).

reknih commented on hayagriva.schema.json:
```diff
@@ -0,0 +1,436 @@
+                        "Klocke, Iny"
+                    ]
+                },
+                "date": {
+                    "type": "string",
+                    "description": "The date at which the item was published.",
+                    "examples": [
+                        "1949-05"
+                    ]
+                },
```
A integer should be allowed here to just specify a year

mkdjr replied:
I'll add `number` as a possible type.

reknih commented on hayagriva.schema.json:
```diff
@@ -0,0 +1,436 @@
+                        }
+                    ]
+                },
+                "serial-number": {
+                    "type": [
+                        "string",
+                        "object"
+                    ],
+                    "description": "Any serial number, including article numbers. If you have serial numbers of well-known schemes like doi, you should put them into the serial number as a dictionary like in the second example. Hayagriva will recognize and specially treat `doi`, `isbn`, `issn`, `pmid`, `pmcid`, and `arxiv`. You can also include `serial` for the serial number when you provide othe ...
+                    "additionalProperties": {
+                        "type": "string"
+                    },
+                    "examples": [
+                        "2003.13722",
+                        {
+                            "doi": "10.22541/au.148771883.35456290",
+                            "arxiv": "1906.00356",
+                            "serial": "8516"
+                        }
+                    ]
+                },
```
Should also support integers

reknih reviewed (commented):
I found a few mismatches with `basic.yml`. I think that before merging this, we should at a test that checks whether `basic.yml` (which is normative) passes the schema. We can do this using [`jsonschema`](https://crates.io/crates/jsonschema).

Freed-Wu commented:
And remember upload it to [schemastore](https://www.schemastore.org/json/)

jassielof commented:
I'd like to contribute to this PR, since I like Hayagriva and am currently working with it for my thesis. How could I help?

PgBiel commented:
> I'd like to contribute to this PR, since I like Hayagriva and am currently working with it for my thesis. How could I help?

We don't have a confirmation yet as to whether these changes will be merged or whether further modifications could be needed, so we can't guarantee a speedy merge, but if you're willing to pick up this work, you're free to address the reviews in your own branch and submit a new PR.

jassielof committed:
Hayagriva JSON Schema

jassielof committed:
done

jassielof committed:
fmt

mkdjr committed:
Merge branch 'typst:main' into main

jassielof committed:
bad title example

jassielof committed:
order

jassielof committed:
title example with single quotes and regex pattern don't support insensitive modifier

jassielof committed:
unrequired title because of parents

jassielof committed:
move examples to defintions and include

mkdjr committed:
Use $id instead of $dynamicAnchor for schema recursion

mkdjr committed:
prevent inclusion of unexpected fields for each entry

mkdjr committed:
make entry type case insensitive

mkdjr committed:
allow `title` field to accept objects as well

mkdjr committed:
allow integers for the `date` field

mkdjr committed:
rearrange fields in `author`

mkdjr committed:
fix `parent` field

mkdjr committed:
specify non-floats for certain fields

mkdjr committed:
allow numbers for serial numbers

mkdjr committed:
fix misspelling of archive

mkdjr committed:
add schema validation and `basic.yml` validation

mkdjr committed:
remove redundant imports

jassielof committed:
removed format date

jassielof committed:
date tilde support

jassielof committed:
no entry type required

jassielof committed:
serial number supports strings or numbers

jassielof committed:
enhance timestamp descriptions and patterns for overflow handling

jassielof committed:
examples moved from any of

jassielof committed:
removed year only string example

jassielof committed:
affiliated list wit unique items

jassielof committed:
arrays with unique items

jassielof committed:
tab size 4 and formatting

jassielof committed:
add schema ID

mkdjr committed:
merge improved JSON schema and reformat `basic.yml`

mkdjr committed:
Merge branch 'jassielof-main'

mkdjr committed:
Edited field descriptions and added examples

mkdjr committed:
disallow additional properties for `formattableString`'s object variant

mkdjr committed:
remove `sentence-case` field from title

jassielof committed:
Entry type is now required in the schema and types are not fully case-insensitive.

Typst `#bibliography()` with a Hayagriva reference doesn't automatically defaults to `Misc` if no type is specified (it asks for a type to be specified).
Aside from that, the types are not fully case-insensitive. If the type is "ArticlE", it'll throw error asking for it to be either "Article" or "article".

jassielof committed:
Affiliated roles are not case insensitive as specified in the file-format.md specifications, but rather all roles are lowercase.

Both commits can be easily reverted in case the Typst validation is fixed as specified in the file-format.md specifications. Otherwise the changes will be kept to comply with the current Typst validation.

gwennlbh commented:
We should probably also mention [`# yaml-language-server: $schema=...`](https://developers.redhat.com/blog/2020/11/25/how-to-configure-yaml-schema-to-make-editing-files-easier#schema_association_without_a_file_name_pattern:~:text=Possibility%202:%20Using%20in%2Dfile%20declarations) and include it in [the examples](https://github.com/typst/hayagriva/blob/main/docs/file-format.md#overview).

It works with the redhat yaml LSP, which I used on nvim before and also works on VS Code (both combined must cover a substential part of Typst users, outside of Typst.app)

jassielof committed:
titles, comments, and examples

there are some inconsistencies with the official docs that have been commented in the schema.
specially against typst

jassielof committed:
Merge branch 'typst:main' into main

jassielof committed:
updated compliance

jassielof commented:
Some updates with the schema and also support for the new chapter field.
I've removed the `$comment` fields thanks to YAML to JSON conversion, although I'm not sure if adding the YAML would be fine.

In any case the YAML version is available in my repository: https://github.com/jassielof/json-schemas/blob/main/docs/hayagriva.yaml

For testing the schema in your Hayagriva YAML, at least with Red Hat's YAML extension in VS Code one can add the following line atop of the file: `# yaml-language-server: $schema=https://raw.githubusercontent.com/mkdjr/hayagriva/refs/heads/main/hayagriva.schema.json`

or using VS Code settings:

```jsonc
  "yaml.schemas": {
    // Replace this with the URL from above
    "https://jassielof.github.io/json-schemas/docs/hayagriva.schema.json": [
      // These for automatic file association to the given schema, these should work for bib.hay, bib.haya, bibliograhy.hay.yml, and such. An official extension by Hayagriva would be good to have as a side note.
      "**/*.{hay,haya,hayagriva}",
      "**/*.{hay,haya,hayagriva}.{yml,yaml}" // Only these will work, since Typst requires the bibliography to have a YAML extension
    ]
  },

```

**Feedback is highly appreciated**, so for any issues, _please comment them here_ so I can solve them.

A related issue to the comments is: <https://github.com/typst/hayagriva/issues/346>.

The schemas (both this and for `typst.toml`) were previously published by me in <https://schemastore.org>, but due to request by the Typst owners are being removed (<https://github.com/SchemaStore/schemastore/pull/5125>, and <https://discord.com/channels/1054443721975922748/1088371867913572452/1437471372438671400>).
